src/transformers/models/univnet/feature_extraction_univnet.py

```diff
@@ -0,0 +1,456 @@
+            waveform,
+            (int((self.n_fft - self.hop_length) / 2), int((self.n_fft - self.hop_length) / 2)),
+            mode="reflect",
+        )
+
+        # Get the complex spectrogram.
+        # Note: waveform must be unbatched currently due to the implementation of spectrogram(...).
+        complex_spectrogram = spectrogram(
+            waveform,
+            window=self.window,
+            frame_length=self.n_fft,
+            hop_length=self.hop_length,
+            fft_length=self.n_fft,
+            power=None,
+            center=self.center,
+            mel_filters=None,
+            mel_floor=None,
+        )
+
+        # Apply the MEL filter bank and MEL floor manually since UnivNet uses a slightly different implementation
+        amplitude_spectrogram = np.sqrt(
+            np.real(complex_spectrogram) ** 2 + np.imag(complex_spectrogram) ** 2 + self.mel_floor
+        )
+        mel_spectrogram = np.matmul(self.mel_filters.T, amplitude_spectrogram)
+
+        # Perform spectral normalization to get the log mel spectrogram.
+        log_mel_spectrogram = np.log(
+            np.clip(mel_spectrogram, a_min=self.compression_clip_val, a_max=None) * self.compression_factor
+        )
+
+        # Return spectrogram with num_mel_bins last
+        return log_mel_spectrogram.T
+
+    def generate_noise(
+        self,
+        noise_length: int,
+        generator: Optional[np.random.Generator] = None,
+    ) -> np.ndarray:
+        """
+        Generates a random noise sequence of standard Gaussian noise for use in the `noise_sequence` argument of
+        [`UnivNetModel.forward`].
+
+        Args:
+            spectrogram_length (`int`):
+                The length (dim 0) of the generated noise.
+            model_in_channels (`int`, *optional*, defaults to `None`):
+                The number of features (dim 1) of the generated noise. This should correspond to the
+                `model_in_channels` of the [`UnivNetGan`] model. If not set, this will default to
+                `self.config.model_in_channels`.
+            generator (`numpy.random.Generator`, *optional*, defaults to `None`)
+                An optional `numpy.random.Generator` random number generator to control noise generation. If not set, a
+                new generator with fresh entropy will be created.
+
+        Returns:
+            `numpy.ndarray`: Array containing random standard Gaussian noise of shape `(noise_length,
+            model_in_channels)`.
+        """
+        if generator is None:
+            generator = np.random.default_rng()
+
+        noise_shape = (noise_length, self.model_in_channels)
+        noise = generator.standard_normal(noise_shape, dtype=np.float32)
+
+        return noise
+
+    def batch_decode(self, waveforms, waveform_lengths=None) -> List[np.ndarray]:
+        r"""
+        Removes padding from generated audio after running [`UnivNetModel.forward`]. This returns a ragged list of 1D
```

Comment:
here the waveform length come from the model that just summed the padding mask returned by this class.
I am guessing this is to have a seamless `batch_decode(model(**input))0`? See my comment regarding the model not needing the masks

Reply:
Yeah, the primary reason is so that 

```python
audio = model(**inputs)
audio = feature_extractor.batch_decode(**audio)
```

works seamlessly. The design is based on [`VitsModel`](https://github.com/huggingface/transformers/blob/e8fdd7875def7be59e2c9b823705fbf003163ea0/src/transformers/models/vits/modeling_vits.py#L1356) and [`VitsModelOutput`](https://github.com/huggingface/transformers/blob/e8fdd7875def7be59e2c9b823705fbf003163ea0/src/transformers/models/vits/modeling_vits.py#L52), although the VITS model is a transformer and thus `VitsModel.forward` uses the `attention_mask` argument in a non-trivial way. See https://github.com/huggingface/transformers/pull/24799#issuecomment-1749833429, https://github.com/huggingface/transformers/pull/24799#issuecomment-1755790002, https://github.com/huggingface/transformers/pull/24799#issuecomment-1755905748, https://github.com/huggingface/transformers/pull/24799#issuecomment-1757395115 for more discussion about the design.

It would be possible to calculate the original waveform lengths in `UnivNetFeatureExtractor` and have `UnivNetModel.forward` take a `waveform_lengths` argument instead, but this also feels weird since `forward` wouldn't do anything with `waveform_lengths` except output it. In my opinion having `forward` accept a `padding_mask` argument feels more natural in `transformers` (although it might be a bit confusing since it's not an attention mask).